src/helpers/render-document-file.js

```diff
@@ -8,41 +8,176 @@ import isVText from 'virtual-dom/vnode/is-vtext';
+  successCount: Array.from(imageCache.values()).filter((v) => v !== null).length,
+  failureCount: Array.from(imageCache.values()).filter((v) => v === null).length,
+  retryStats,
+});
+
+// Helper function for conditional verbose logging
+const logVerbose = (verboseLogging, message, ...args) => {
+  if (verboseLogging) {
+    // eslint-disable-next-line no-console
+    console.log(message, ...args);
+  }
+};
+
+
 // eslint-disable-next-line consistent-return, no-shadow
-export const buildImage = async (docxDocumentInstance, vNode, maximumWidth = null) => {
+export const buildImage = async (
+  docxDocumentInstance,
+  vNode,
+  maximumWidth = null,
+  options = {}
+) => {
+  // Extract image processing options with defaults from constants.js
+  const maxRetries =
+    options.maxRetries ||
+    docxDocumentInstance.imageProcessing?.maxRetries ||
+    defaultDocumentOptions.imageProcessing.maxRetries;
+  const verboseLogging =
+    options.verboseLogging ||
+    docxDocumentInstance.imageProcessing?.verboseLogging ||
+    defaultDocumentOptions.imageProcessing.verboseLogging;
   let response = null;
   let base64Uri = null;
+
   try {
     const imageSource = vNode.properties.src;
-    if (isValidUrl(imageSource)) {
-      const base64String = await imageToBase64(imageSource).catch((error) => {
-        // eslint-disable-next-line no-console
-        console.warn(`skipping image download and conversion due to ${error}`);
+
+    // Check cache first for external URLs
+    if (isValidUrl(imageSource) && imageCache.has(imageSource)) {
+      const cachedData = imageCache.get(imageSource);
+      if (cachedData === null) {
+        // Previously failed to download in this document generation, skip this image
+        logVerbose(
+          verboseLogging,
+          `[CACHE] Skipping previously failed image in this document: ${imageSource}`
+        );
         return null;
-      });
+      }
+      logVerbose(verboseLogging, `[CACHE] Using cached image data for: ${imageSource}`);
+      base64Uri = cachedData;
+      // Update vNode to reflect the cached data URL for subsequent processing
+      vNode.properties.src = base64Uri;
+    } else if (isValidUrl(imageSource)) {
+      // Download and cache the image with retry mechanism
+      let base64String = null;
+      let lastError = null;
+
+      for (let attempt = 1; attempt <= maxRetries; attempt += 1) {
+        retryStats.totalAttempts += 1;
+
+        try {
+          logVerbose(
+            verboseLogging,
+            `[RETRY] Attempt ${attempt}/${maxRetries} for: ${imageSource}`
+          );
+
+          // Use configurable timeout, default 5 seconds, with exponential backoff for retries
+          const baseTimeout = Math.max(1000, Math.min(options.downloadTimeout || 5000, 30000)); // 1s-30s range
+          const timeoutMs = baseTimeout * attempt;
+          const maxSizeBytes = Math.max(1024, options.maxImageSize || 10 * 1024 * 1024); // min 1KB
```

Comment:
Let's save these values in a constants, instead of hardcoding them here, and use.